source/beamline/PGM/PGMBPMControl.cpp

```diff
@@ -0,0 +1,38 @@
+#include "beamline/PGM/PGMBPMControl.h"
+
+PGMBPMControl::PGMBPMControl(const QString &name, const QString &pvName, int avgValue, int variance, QObject *parent)
+    : QObject(parent)
+
+{
+    currentValue_ = 0;
+    averageValue_ = avgValue;
+    variance_ = variance;
+    state_ = true;
+    bpmPV_ = new AMReadOnlyPVControl(name, pvName, this);
+
+    // Get updated pv value from AMReadOnlyPVControl.
+    connect(bpmPV_, SIGNAL(valueChanged(double)), this, SLOT(setCurrentValue(double)));
+
+}
+
+void PGMBPMControl::updateValidity(double newValue){
+
+    int upperBound = averageValue_ + variance_;
+    int lowerBound = averageValue_ - variance_;
+
+    bool newState = (newValue < upperBound || newValue > lowerBound);
+
+    if (newState != state_){
+
+        state_ = newState;
+        emit stateValidityChanged(state_);
+    }
+}
+
+void PGMBPMControl::setCurrentValue(double value){
+
+    currentValue_ = value;
+    updateValidity(currentValue_);
+    emit onBPMValueChanged(QString("%1 um").arg(value, 0, 'f', 0));
```

Comment:
Things associated with display should be in the view regardless of which view it is. Conversely, you could have a BPM view that does this for you.

Sent from my iPhone

On Apr 27, 2016, at 10:19 PM, David Leacock <notifications@github.com<mailto:notifications@github.com>> wrote:


In source/beamline/PGM/PGMBPMControl.cpp<https://github.com/acquaman/acquaman/pull/2213#discussion_r61371842>:

> +    int lowerBound = averageValue_ - variance_;
> +
> +    bool newState = (newValue < upperBound || newValue > lowerBound);
> +
> +    if (newState != state_){
> +
> +        state_ = newState;
> +        emit stateValidityChanged(state_);
> +    }
> +}
> +
> +void PGMBPMControl::setCurrentValue(double value){
> +
> +    currentValue_ = value;
> +    updateValidity(currentValue_);
> +    emit onBPMValueChanged(QString("%1 um").arg(value, 0, 'f', 0));


Got it. My thinking was to keep as much out of the persistent view as possible. Would you like me to change it?

—
You are receiving this because you commented.
Reply to this email directly or view it on GitHub<https://github.com/acquaman/acquaman/pull/2213/files/5421a984a5ef75f7bf7b4ae66dffda8135251551..87b0103c1ec18f25dfdff86ed07842babb0b30dd#r61371842>